pghoard/rohmu/delta/snapshot.py

```diff
@@ -0,0 +1,231 @@
+# Copyright (c) 2021 Aiven, Helsinki, Finland. https://aiven.io/
+import base64
+import logging
+import os
+import threading
+from pathlib import Path
+from typing import Callable, Optional
+
+from pghoard.rohmu.delta.common import (
+    EMBEDDED_FILE_SIZE, Progress, SnapshotFile, SnapshotHash, SnapshotState, hash_hexdigest_readable,
+    increase_worth_reporting, parallel_map_to
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Snapshotter:
+    """Snapshotter keeps track of files on disk, and their hashes.
+
+        The hash on disk MAY change, which may require subsequent
+        incremential snapshot and-or ignoring the files which have changed.
+
+        The output to outside is just root object's hash, as well as list
+        of other hashes which correspond to files referred to within the
+        file list contained in root object.
+
+        Note that any call to public API MUST be made with
+        snapshotter.lock held. This is because Snapshotter is process-wide
```

Comment:
One option protecting against calling these methods w/o preceding `lock` being held by making these methods protected and introducing method creating context manager locking `lock` and returning facade object exposing these 3 methods.

Reply:
As PR description says, this part is taken from Astacus, I did not want to change the behavior and interfaces, so it can be re-used as is.

Reply:
Is it possible to split the part taken from Astacus into a separate commit?

Reply:
I could, but it's pretty easy to split it, everything in rohmu folder is taken from Astacus almost as it is.

Reply:
Sorry for me being out of the context: I wonder is it possible to split this functionality shared by Astacus and pghoard into a separate library?

Reply:
This is what's probably coming, was discussed multiple times, rohmu should probably move to a separate library itself.

Reply:
Sounds like a good time to create a story :wink:

Reply:
Yes, but that's a bit out of the scope of the current ticket :) and will require quite some effort as touches many places.

Reply:
Can you create a story? :blush:

Reply:
yep, will check first maybe we have something already.

Reply:
Ticket created.